proto/buf/compiler/v1alpha1/ast.proto

```diff
@@ -0,0 +1,356 @@
+// Copyright 2020-2024 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+syntax = "proto3";
+
+package buf.compiler.v1alpha1;
+
+import "buf/compiler/v1alpha1/report.proto";
+
+// NOTE: In this file, we number all Span-typed fields together, starting
+// at tag 10. They must also be suffixed with `_span`.
+
+// A parsed AST file. This is the root file for the whole Protocompile AST.
+message File {
+  // The original filesystem file this file was parsed from.
+  Report.File file = 1;
+  // Declararations in this file.
+  repeated Decl decls = 2;
+}
+
+// A source code span for a specific `File`.
+//
+// This only contains byte offsets for the span; all other information
+// (such as the line number) should be re-computed as needed.
+message Span {
+  uint32 start = 1;
+  uint32 end = 2;
+}
+
+// A path in a Protobuf file.
+//
+// Protobuf paths are quite complex: they are not just dot-separated squences
+// of names: each component can itself be a Protobuf path, such as
+// `foo.(bar.(baz)).zing`. Although deep nesting is not currently used in
+// the language, this AST supports it.
+message Path {
+  // A path component.
+  message Component {
+    enum Separator {
+      SEPARATOR_UNSPECIFIED = 0;
+      SEPARATOR_DOT = 1;
+      SEPARATOR_SLASH = 2;
```

Comment:
Well, the thing is that this is an AST, not something you would be expected to do the semantic analysis of URLs on. This needs to very faithfully capture what is generated by the parser for the purpose of testing.

When we start building the IR module, that will have a representation that respects this semantic restriction.

Regarding `[...]`, the way this is represented in the AST is as an `ExprField` with a one-element `ExprArray` key. This may seem a bit silly, but reinterpreting the expression in the key side of an `ExprField` later, after all parsing is complete, ultimately results in a simpler parser (this is similar to how when parsing a def, we don't really care what precisely it is until parsing of the whole file is complete).

The reason for doing it this way is because, if a user writes `{ [xyz] }`, it is not clear as we are entering the `[...]` whether this is going to be a misplaced array or an extension name.

Reply:
> Well, the thing is that this is an AST, not something you would be expected to do the semantic analysis of URLs on.

I'm talking about _syntactic_ analysis of URLs. The parser rejects a rather significant portion of valid URLs here because its productions only recognize paths with slash separators. While the new experimental parser is more lenient (and allows arbitrary path elements, unlike protoc or the current protocompile parser), it still doesn't allow a scheme prefix (i.e. "https://") and doesn't allow domain or path components unless they match the syntax of a protobuf identifier (so they'd reject domain components with a hyphen, hostnames that refer to IP addresses instead of domain names, and path components that include punctuation or percent-encoded characters).

In any event, no action needed here I guess. This was me thinking out loud and wondering if a TODO in the relevant part of the AST (and maybe in the parser, too) would be useful...

Reply:
I don't think a TODO is needed. We're going to run into this problem organically and I'd rather deal with it when it comes up.